extensions-contrib/moving-average-query/src/main/java/org/apache/druid/query/movingaverage/RowBucketIterable.java

```diff
@@ -0,0 +1,147 @@
+package org.apache.druid.query.movingaverage;
+
+import org.apache.druid.data.input.Row;
+import org.apache.druid.java.util.common.guava.Sequence;
+import org.apache.druid.java.util.common.guava.Yielder;
+import org.joda.time.DateTime;
+import org.joda.time.Interval;
+import org.joda.time.Period;
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+/**
+ * It is the iterable used to bucket data into days,
+ * doing appropriate lookahead to see if the next row is in the same day or a new day.
+ */
+public class RowBucketIterable implements Iterable<RowBucket>
+{
+
+  public final Sequence<Row> seq;
+  private List<Interval> intervals;
+  private Period period;
+
+  public RowBucketIterable(Sequence<Row> seq, List<Interval> intervals, Period period)
+  {
+    this.seq = seq;
+    this.period = period;
+    this.intervals = intervals;
+  }
+
+  /* (non-Javadoc)
+   * @see java.lang.Iterable#iterator()
+   */
+  @Override
+  public Iterator<RowBucket> iterator()
+  {
+    return new RowBucketIterator(seq, intervals, period);
+  }
+
+  static class RowBucketIterator implements Iterator<RowBucket>
+  {
+    private Yielder<RowBucket> yielder;
+    private DateTime endTime;
+    private DateTime expectedBucket;
+    private Period period;
+    private int intervalIndex = 0;
+    private List<Interval> intervals;
+    private boolean processedLastRow = false;
+    private boolean processedExtraRow = false;
+
+    public RowBucketIterator(Sequence<Row> rows, List<Interval> intervals, Period period)
+    {
+      this.period = period;
+      this.intervals = intervals;
+      expectedBucket = intervals.get(intervalIndex).getStart();
+      endTime = intervals.get(intervals.size() - 1).getEnd();
+      yielder = rows.toYielder(null, new BucketingAccumulator());
+    }
+
+    /* (non-Javadoc)
+     * @see java.util.Iterator#hasNext()
+     */
+    @Override
+    public boolean hasNext()
+    {
+      return expectedBucket.compareTo(endTime) < 0;
+    }
+
+    /* (non-Javadoc)
+     * @see java.util.Iterator#next()
+     */
+    @Override
+    public RowBucket next()
+    {
+      RowBucket currentBucket = yielder.get();
+
+      if (expectedBucket.compareTo(intervals.get(intervalIndex).getEnd()) >= 0) {
+        intervalIndex++;
+        if (intervalIndex < intervals.size()) {
+          expectedBucket = intervals.get(intervalIndex).getStart();
+        }
+      }
+      // currentBucket > expectedBucket
+      if (currentBucket != null && currentBucket.getDateTime().compareTo(expectedBucket) > 0) {
+        currentBucket = new RowBucket(expectedBucket, Collections.emptyList());
+        expectedBucket = expectedBucket.plus(period);
+        return currentBucket;
+      }
+
+      if (!yielder.isDone()) {
+        // standard case. return regular row
+        yielder = yielder.next(currentBucket);
+        expectedBucket = expectedBucket.plus(period);
+        return currentBucket;
```

Comment:
This might not return the very last rowBucket. For the last rowBucket, the updated `yielder` should hold it, but `hasNext()` would return false if `expectedBucket >= endTime`.

Reply:
`hasNext()` is correct since interval upper bound is open-ended.

Reply:
Hmm,  let me add some more details. `yielder` is updated in this `if` clause which should be used to iterate all values in it. However, in `hasNext()`, it only checks `expectedBucket` is less than `endTime`. Since `expectedBucket` is also updated in this `if` clause, `hasNext()` can return false even though `yielder` is not used yet.